src/V3DfgDfgToAst.cpp

```diff
@@ -0,0 +1,318 @@
+            // create a temporary
+            if (const DfgVar* const thisDfgVarp = vtxp->cast<DfgVar>()) {
+                // This is a DfgVar
+                varp = getCanonicalVar(thisDfgVarp);
+            } else if (const DfgVar* const sinkDfgVarp = vtxp->findSink<DfgVar>()) {
+                // We found a DfgVar driven by this node
+                varp = getCanonicalVar(sinkDfgVarp);
+            } else {
+                // No DfgVar driven by this node. Create a temporary.
+                const string name = m_tmpNames.get(vtxp->hash(m_hashCache).toString());
+                // TODO: Do we ever need to make this VSigning::SIGNED?
+                AstNodeDType* const dtypep = v3Global.rootp()->findBitDType(
+                    vtxp->width(), vtxp->width(), VSigning::UNSIGNED);
+                varp = new AstVar{vtxp->fileline(), VVarType::MODULETEMP, name, dtypep};
+                // Add temporary AstVar to containing module
+                m_modp->addStmtp(varp);
+            }
+            // Add to map
+        }
+        return varp;
+    }
+
+    AstNodeMath* convertDfgVertexToAstNodeMath(DfgVertex* vtxp) {
+        UASSERT_OBJ(!m_resultp, vtxp, "Result already computed");
+        iterate(vtxp);
+        UASSERT_OBJ(m_resultp, vtxp, "Missing result");
+        AstNodeMath* const resultp = m_resultp;
+        m_resultp = nullptr;
+        return resultp;
+    }
+
+    AstNodeMath* convertSource(DfgVertex* vtxp) {
+        if (vtxp->hasMultipleSinks()) {
+            // Vertices with multiple sinks need a temporary variable, just return a reference
+            return new AstVarRef{vtxp->fileline(), getResultVar(vtxp), VAccess::READ};
+        } else {
+            // Vertex with single sink is simply recursively converted
+            UASSERT_OBJ(vtxp->hasSinks(), vtxp, "Must have one sink: " << vtxp->typeName());
+            return convertDfgVertexToAstNodeMath(vtxp);
+        }
+    }
+
+    // VISITORS
+    void visit(DfgVertex* vtxp) override {  // LCOV_EXCL_START
+        vtxp->v3fatal("Unhandled DfgVertex: " << vtxp->typeName());
+    }  // LCOV_EXCL_STOP
+
+    void visit(DfgVar* vtxp) override {
+        m_resultp = new AstVarRef{vtxp->fileline(), getCanonicalVar(vtxp), VAccess::READ};
+    }
+
+    void visit(DfgConst* vtxp) override {  //
+        m_resultp = vtxp->constp()->cloneTree(false);
+    }
+
+    // The rest of the 'visit' methods are generated by 'astgen'
+#include "V3Dfg__gen_dfg_to_ast.h"
+
+    // Predicate checking whether a variable should be retained,
+    // even if it's equal to another variable
+    static bool keep(const DfgVar* vtxp) {
+        // Keep if referenced outide this module
+        if (vtxp->hasExtRefs()) return true;
+        // Keep if traced
+        if (v3Global.opt.trace() && vtxp->varp()->isTrace()) return true;
+        // Otherwise it can be removed
+        return false;
+    }
```

Comment:
Generally we throw away redundant variables. This tells us when not to. Can you think of any more cases? Coverage is inserted before DFG, so I think that should be fine.

Reply:
public.

Reply:
Fixed